### DISCUSSION TITLE
Reverse proxy from host to Docker (using nginx proxy_pass) - clarification

### DISCUSSION THREAD

**Discussion by bytebraid** [AzuraCast/AzuraCast#3882] [Q&A]
Firstly, what a great piece of work Azuracast is, thanks so much for it.

Now a slight technical challenge I've spent some hours trying to get a grip on docker-proxy and what it actually does, I'm still not sure of either and am seeking some brief clarification on the matter. 

Desired configuration:

1. Azuracast on port 7777/7778 (inside docker's oblique network configuration)
2. Nginx on port 80 running outside docker

I understand the current way of solving this is crow-barring the host nginx config inside azuracast as per https://www.azuracast.com/extending/multisite.html

I understand also that this is what you support officially, so I'm not making a feature request.

All I'm asking is for some aid in understanding docker's network-level config. 

With this nginx config:

location /radio {
   proxy_pass http://localhost:7777/public/radio
}

It simply doesn't work (the request hangs and eventually a TCP RST appears, nginx serves 502 Bad Gateway) - likely as the kernel doesn't permit NAT routing on loopback

So looking at the iptables packet counts, nothing makes it to the DNAT targets in the prerouting table.

strace on the relevant docker-proxy bound to 7777 shows no legwork being done in that process, so I assume the packet is bouncing around iptables until the RST comes from a timeout in the kernel stack.. so azuracast makes no use of docker-proxy.

However, Is it possible to coax this kind of config to work by communicating on a non-loopback interface? 

I'm wondering if anyone has had success with this kind of setup.

**Comment by Vaalyn** [comment]
The easiest way to change the port that AzuraCast is running on would be to use the `./docker.sh change-ports` command and change the ports `80` and `443` to the ones you'd like it to use.

**Comment by bytebraid** [comment]
Not quite what I'm asking, the ports in use are desired, I would just like to understand docker's packet routing, in order to front the azuracast installation with an nginx proxy running in userland on the host

**Comment by Vaalyn** [comment]
The `web` container is the one running the AzuraCast PHP application, internally also using an NGINX. This container is part of the `frontend` Docker network and is not externally reachable from the outside of the Docker network. The `nginx_proxy` is also part of the `frontend` Docker network and is externally reachable since it has its ports mapped to the outside/host. In order to be reachable by something outside of a Docker network the container must have its ports mapped to ports on the host.

If you want to use your own proxy, either do what I wrote to change the default ports for the `nginx_proxy` to something else and then use your own proxy to connect to these new ports of the `nginx_proxy` or remove the NGINX proxy as well as the `nginx_proxy_letsencrypt` and forward the ports of the web container to the outside.

The second option is not officially support by us so we will not be providing assistance in getting this working or debugging problems with it. You are free to do this and we do encourage you to share your progress with this though.

**Comment by Cedric-Demarais** [comment]
Hi

I do reverse proxy Azuracast with Apache 2.4.

You need to enable these mods: proxy / proxy html / headers / rewrite (for http redirection to https)

This vhost configuration is set for domain name `radio.myradio.com`.

AzuraCast Web HTTP Port is set to `10080` via `./docker.sh change-ports`

In my setup, i redirect any HTTP request on that domain to HTTPS.

radio.myradio.com.conf:

\<VirtualHost *:80>


        ServerName radio.myradio.com
        ErrorLog ${APACHE_LOG_DIR}/error-radio.myradio.com.log
        CustomLog ${APACHE_LOG_DIR}/access-radio.myradio.com.log combined
        DocumentRoot /www/radio.myradio.com

        Options Indexes FollowSymLinks MultiViews
        ErrorDocument 404 /index.html
        ErrorDocument 403 /index.html
        ProxyPreserveHost On
        ServerAdmin webmaster@localhost
        LogLevel warn
        ServerSignature Off

        RewriteEngine on
        RewriteCond %{SERVER_NAME} =radio.myradio.com
        RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]


\</VirtualHost>



radio.myradio.com-le-ssl.conf:

\<IfModule mod_ssl.c>
\<VirtualHost *:443>

        ServerName radio.myradio.com
        ErrorLog ${APACHE_LOG_DIR}/error-radio.myradio.com.log
        CustomLog ${APACHE_LOG_DIR}/access-radio.myradio.com.log combined
        DocumentRoot /www/radio.myradio.com
        LogLevel warn
        ServerSignature Off
        ServerAdmin webmaster@localhost
        ErrorDocument 404 /
        ErrorDocument 403 /
        Options Indexes FollowSymLinks MultiViews
        ProxyRequests Off

        ProxyPreserveHost On
        ProxyPass / http://192.168.1.250:10080/
        ProxyPassReverse / http://192.168.1.250:10080/
        ProxyHTMLEnable On
        ProxyHTMLURLMap http://192.168.1.250:10080/ /
        ProxyHTMLURLMap / /
        ProxyHTMLCharsetOut *
        RequestHeader unset Accept-Encoding

SSLCertificateFile /etc/letsencrypt/live/radio.myradio.com/fullchain.pem
SSLCertificateKeyFile /etc/letsencrypt/live/radio.myradio.com/privkey.pem
Include /etc/letsencrypt/options-ssl-apache.conf

\</VirtualHost>
\</IfModule>

Note: Docker running on synology, maybe you don't need these:
`        ProxyHTMLCharsetOut *
        RequestHeader unset Accept-Encoding`

Note2: I think this line is useless too, but it's in my reverse proxy configuration template:
`ProxyHTMLURLMap / /`
